### ISSUE TITLE
Service does not start on boot when app is installed to external storage

### ISSUE THREAD

**Issue by basileus-sys** [biqqles/p2oggle#5]
After a reboot the service that monitors the switch does not automatically start requiring the user to manually open the app. Once done that the app notifies "SU rights granted" and starts working. Only after that P2oggle appears as an entry on the "running services" section of dev options.

Tested on a fresh install of: 

LineageOS 17.1 20200927-nightly
Open gapps arm64-10.0-pico-20201003
Magisk 20.4 + Magisk manager 8.0.0 (302)

**Comment by biqqles** [comment]
Sorry for the late reply. I evidently need to fiddle with my GitHub notification settings.

Thanks for taking the time to report this. This is the first report of the service failing to start that I'm aware of. When did the app work last for you? Lineage 16/17 or an older build of 17.1?

**Comment by basileus-sys** [comment]
No worries (same here it looks like lol)

I have bought my P2 only recently so i have never tried past versions of LineageOS so I can't tell if in past builds it worked or not.

**Comment by biqqles** [comment]
OK, that's a pity (welcome to the P2 community though!). I'm at a loss for an explanation to be honest. Could it be battery saver mode perhaps? Or do other services (e.g. Google Play Services) start upon boot?

There's also a chance that the service is somehow immediately crashing or otherwise failing to start when launched from outside the app's activity. You could [take a logcat](https://www.xda-developers.com/guide-sending-a-logcat-to-help-debug-your-favorite-app/) just after boot and see if P2oggle emits anything.

**Comment by basileus-sys** [comment]
Thank you a lot, the P2 is great no fuss phone and im greatly enjoying it!

Great idea to suggest the logcat monitor, i have rebooted the phone and checked the logs and there were a couple of entries saying stuff along the lines of "p2oggle not found, stopping" so I got suspicious because I have a microSD installed (in the "extend internal storage" format) so i checked if P2oggle was installed on external storage and it was, so i moved it to internal, rebooted and now it works perfectly!! Problem solved.

**Comment by biqqles** [comment]
Interesting, did you move the app to/from external storage in Settings? If so I can probably disable that from being possible now we know it causes issues.

Anyway, glad we figured that out!

**Comment by basileus-sys** [comment]
Yes, I went into the app info>storage&cache>change storage

**Comment by biqqles** [comment]
Perfect, thanks. I will get that fixed.

**Comment by biqqles** [comment]
Well now I'm confused. The Android docs [say](https://developer.android.com/guide/topics/manifest/manifest-element.html) this:

> The user may also request to move an app from the internal storage to the external storage. However, the system will not allow the user to move the app to external storage if this attribute is set to `internalOnly`, which is the default setting.

And since I have left it at the default it shouldn't have been possible to install the app to external storage in the first place. Perhaps Lineage ignores this attribute. Or do you have "Force allow apps on external" enabled in Developer options?

**Comment by basileus-sys** [comment]
Everything should be in the default configuration, i havent touched anything regarding that. I inserted my microsd, formatted it and let the phone arbitrarily decide wheater to place new apps in the internal memory or external. I dont really understand how it decides what goes were, on app installation there is no such thing as a "select storage" prompt, it just.... does its thing.

Right now (23:31) my P2 is teared open since tomorrow morning im picking up a package at the post office with a new home button (fp scanner doesnt work properly) i should have the phone back online in 9-11 hrs, then i'll double check if the "Force allow apps on external" option is enabled and let you know

**Comment by basileus-sys** [comment]
> 
> 
> Interesting, did you move the app to/from external storage in Settings? If so I can probably disable that from being possible now we know it causes issues.
> 
> Anyway, glad we figured that out!

Oh wait wait wait, i double read everything and i think there have been a misunderstanding: i did move the app in settings, **from external to internal to fix it**, thats what i meant. I confirm that the option to move p2oggle to external is missing, after moving it from external to internal i no longer could move the app back to external

**Comment by basileus-sys** [comment]
Now that i think about it i think that once i realized that there were problems with P2oggle not starting i tried fiddling a bit formatting the phone and reinstalling lineage+opengapps and p2oggle directly through TWRP using a usb otg stick. I dont think that TWRP could have installed p2oggle on external since after the formatting it shouldnt have recognized the microsd without being formatted again first, so i guess the most likely culprit is lineageos moving p2oggle to external when formatting the microsd (?)

**Comment by basileus-sys** [comment]
Im back, I confirm "Force allow apps on external" was disabled.

**Comment by biqqles** [comment]
OK, thanks for all the info. It looks like Lineage (or likely upstream AOSP) unhelpfully ignores the storage preferences of apps when formatting a new SD card. Since the bug is effectively with that process instead of this app I'll mark this as wontfix and close.
